### PR TITLE
Add array_duplicates Presto function

### PR DESCRIPTION
**Pull request by darrenfu**
## Summary
* Adding the array_duplicates Presto function to Velox

`array_duplicates(array(T))`

> Returns a set of elements that occur more than once in array. 
> T must be coercible to bigint or varchar.

For example:
```
select ARRAY_DUPLICATES(ARRAY [5, 2, 5, 1, 1, 5, null, null])); -- [null, 1, 5]
select ARRAY_DUPLICATES(ARRAY [true, true])); -- (SYNTAX_ERROR) line 1:8: Unexpected parameters (array(boolean)) for function array_dupes. Expected: array_duplicates(array(varchar)) , array_duplicates(array(bigint))
```

Diff: D31718890

### CODE REVIEW

File: velox/functions/prestosql/ArrayDuplicates.cpp [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+              uniqueMap[value] = false;
+            }
+          } else {
+            uniqueMap[value] = true;
+          }
+        }
+      }
+
+      uniqueMap.clear();
+      rawSizes[row] = indicesCursor - rawOffsets[row];
```
Comment (darrenfu) [inline velox/functions/prestosql/ArrayDuplicates.cpp:121]: Per @mbasmanova's advice, fixed an existing non-contiguous row bug via pointer increment.

File: velox/functions/prestosql/tests/ArrayDuplicatesTest.cpp [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+      {S("blue clear sky above")},
+      {std::nullopt},
+      {},
+      {},
+  });
+
+  testExpr(expected, "array_duplicates(C0)", {array});
+}
+
+TEST_F(ArrayDuplicatesTest, testNonContiguousRows) {
```
Comment (darrenfu) [inline velox/functions/prestosql/tests/ArrayDuplicatesTest.cpp:164]: Thanks for @mbasmanova's new test case to verify the fix on `rawSizes` and `rawOffsets` pointers above.

File: velox/functions/prestosql/ArrayDuplicates.cpp [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+      vector_size_t numOfNulls = 0;
+      for (vector_size_t i = offset; i < offset + size; ++i) {
+        if (elements->isNullAt(i)) {
+          numOfNulls++;
+          if (numOfNulls == 2) {
+            rawNewIndices[indicesCursor++] = i;
+          }
+        } else {
+          auto value = elements->valueAt<T>(i);
+          if (uniqueMap.contains(value)) {
```
Comment (darrenfu) [inline velox/functions/prestosql/ArrayDuplicates.cpp:109]: I still feel we should use F14Map here instead of `F14Set.insert(value).second` because we need to find all "duplicates" instead of "unique" values. The bool flag in the Map is to indicate whether the value occurs once or more than once.
cc @mbasmanova

File: velox/functions/prestosql/ArrayDuplicates.cpp [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+  ///
+  /// In order to prevent copies of array elements, the function reuses the
+  /// internal elements() vector from the original ArrayVector.
+  ///
+  /// First a new vector is created containing the indices of the elements
+  /// which will be present in the output, and wrapped into a DictionaryVector.
+  /// Next the `lengths` and `offsets` vectors that control where output arrays
+  /// start and end are wrapped into the output ArrayVector.
+
+  ArrayDuplicatesFunction() {}
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayDuplicates.cpp:46]: this is not needed; the compiler will create a default constructor automatically

File: velox/functions/prestosql/ArrayDuplicates.cpp [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+  // Validate element type inside the vector
+  void validateElementType(const TypeKind& kind) const {
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayDuplicates.cpp:49]: this function can be private and static

File: velox/functions/prestosql/ArrayDuplicates.cpp (hunk 3 above)
Comment (mbasmanova) [reply velox/functions/prestosql/ArrayDuplicates.cpp:109]: I think I understand this. Sounds fine. Thanks for clarifying.

Comment (mbasmanova) [reply velox/functions/prestosql/ArrayDuplicates.cpp:109]: nit: you may want to reduce the number of lookups in the map by using uniqueMap.find() which returns an iterator to the entry and can be used to check whether the entry exists as well as what is the value.

File: velox/functions/prestosql/ArrayDuplicates.cpp [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+            uniqueMap[value] = true;
+          }
+        }
+      }
+
+      uniqueMap.clear();
+      rawSizes[row] = indicesCursor - rawOffsets[row];
+    });
+
+    newIndices->setSize(indicesCursor * sizeof(vector_size_t));
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayDuplicates.cpp:124]: nit: not for this PR, but it would be nice to change this method to a template and rename it to resize:

```
newIndices->resize<vector_size_t>(indicesCursor);
```

File: velox/functions/prestosql/tests/ArrayDuplicatesTest.cpp (hunk 2 above)
Comment (mbasmanova) [reply velox/functions/prestosql/tests/ArrayDuplicatesTest.cpp:164]: naming: testNonContiguousRows -> nonContiguousRows

**Review by mbasmanova — COMMENTED** [review]
@darrenfu Looks good to me % minor comments below. It would be nice to add e2e test to TestHiveQueries.java .

**Comment by darrenfu** [comment]
Update: 
1. Implement sort
2. [optimization] Use `uniqueMap.find()` to minimize visiting the map
3. Refine test cases

File: velox/functions/prestosql/ArrayDuplicates.cpp [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+          auto it = uniqueMap.find(value);
+          if (it == uniqueMap.end()) {
+            uniqueMap[value] = true;
+          } else if (it->second) {
+            it->second = false;
+            rawNewIndices[indicesCursor] = i;
+            // [Sort] if it is the first non-null element, do nothing;
+            // otherwise, swap with the previous non-null element if the current
+            // is less
+            if (!((indicesCursor == rawOffsets[row] + 1 && numOfNulls > 1) ||
```
[the remark this answers is not in the document]
Comment (darrenfu) [reply velox/functions/prestosql/ArrayDuplicates.cpp:119]: @mbasmanova I understand the sorting logic is a bit hard to understand even with inline comments.
Yes, this new logic passed the following two unit test cases:


I also added a section "Sorting Algorithm Explanation" in this PR to explain more details. The time complexity will be O(N) because of only one-time scan, which is better than std::sort() (O(N*Log(N)) on the duplicate result array.

[the remark this answers is not in the document]
Comment (darrenfu) [reply velox/functions/prestosql/ArrayDuplicates.cpp:119]: I wanna achieve sorting in the original one-time scan. Looks like I missed one corner case when insert a non-null value. Let me revise the logic.

File: velox/functions/prestosql/ArrayDuplicates.cpp [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+  validateType(inputArgs);
+  auto elementType = inputArgs.front().type->childAt(0);
+
+  return VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
+      createTyped, elementType->kind(), inputArgs);
+}
+
+// Define function signature.
+std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+  // array(T) -> array(T)
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayDuplicates.cpp:201]: I just noticed that this signature is too broad. It says that the function accepts any array(T), but this function allows only a subset of T, e.g. tinyint, smallint, integer, bigint and varchar. Please, update.

File: velox/functions/prestosql/ArrayDuplicates.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (darrenfu) [reply velox/functions/prestosql/ArrayDuplicates.cpp]: Take my word back, reimplement it with std::sort plus a nullable sort comparator.

File: velox/functions/prestosql/ArrayDuplicates.cpp [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+
+      std::sort(
+          rawNewIndices + rawOffsets[row],
+          rawNewIndices + indicesCursor,
+          sparksql::Index2ValueNullableLess<T>(elements));
+    });
+
+    newIndices->setSize(indicesCursor * sizeof(vector_size_t));
+    auto newElements =
+        BaseVector::transpose(newIndices, std::move(elementsVector));
```
Comment (darrenfu) [inline velox/functions/prestosql/ArrayDuplicates.cpp:123]: If the input type is: `array(tinyint/smallint/int)`, the output type should be `array(bigint)`.

Do I need to explicitly create a new int64_t `ArrayVector` and copy all the values from `elementsVector`? If yes, is it possible to achieve that by borrowing the idea of `BaseVector.copy()` + `ArrayVector.copy()`?

Comment (darrenfu) [reply velox/functions/prestosql/ArrayDuplicates.cpp]: nvm, figured it out in c53da2c9d9fcdd29cec8a496db686cbbeb0d4010

**Comment by mbasmanova** [comment]
@darrenfu GitHub is having some trouble displaying PRs will many commits. Any chance you could squash all commits into one and update the PR?

**Comment by mbasmanova** [comment]
@darrenfu There are build failures. Would you take a look?

File: velox/docs/functions/coverage.rst [hunk 10]
```diff
@@ -135,10 +135,10 @@ Here is a list of all scalar and aggregate Presto functions with functions that
     any_match                                 date_add                                  is_json_scalar                            repeat                                    st_relate                                     approx_set
     array_average                             date_diff                                 is_nan                                    :func:`replace`                           st_startpoint                                 :func:`arbitrary`
     :func:`array_distinct`                    date_format                               is_subnet_of                              :func:`reverse`                           st_symdifference                              :func:`array_agg`
-    array_dupes                               date_parse                                jaccard_index                             rgb                                       st_touches                                    :func:`avg`
+    :func:`array_duplicates`                  date_parse                                jaccard_index                             rgb                                       st_touches                                    :func:`avg`
```
Comment (mbasmanova) [inline velox/docs/functions/coverage.rst:138]: Thank you for noticing that coverage maps need updating as well. We update these in bulk using the procedure described in https://github.com/facebookincubator/velox/tree/main/velox/functions/prestosql/coverage#readme and submit changes as separate PRs, e.g. #454 . Hence, let's not update these here.

File: velox/functions/prestosql/ArrayDuplicates.cpp [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+        toElementRows(elementsVector->size(), rows, arrayVector);
+    exec::LocalDecodedVector elements(context, *elementsVector, elementsRows);
+
+    vector_size_t elementsCount = elementsRows.size();
+    vector_size_t rowCount = arrayVector->size();
+
+    // Allocate new vectors for indices, length and offsets.
+    memory::MemoryPool* pool = context->pool();
+    BufferPtr newIndices =
+        AlignedBuffer::allocate<vector_size_t>(elementsCount, pool);
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayDuplicates.cpp:69]: use newly introduced helper functions: allocateIndices/Offsets/Sizes. See #464 and #465 for context.

File: velox/functions/prestosql/ArrayDuplicates.cpp [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+    BufferPtr newOffsets =
+        AlignedBuffer::allocate<vector_size_t>(rowCount, pool);
+
+    // Pointers and cursors to the raw data.
+    vector_size_t indicesCursor = 0;
+    auto* rawNewIndices = newIndices->asMutable<vector_size_t>();
+    auto* rawSizes = newLengths->asMutable<vector_size_t>();
+    auto* rawOffsets = newOffsets->asMutable<vector_size_t>();
+
+    // Convert coercible-numeric-typed ArrayVector to bigint ArrayVector
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayDuplicates.cpp:81]: I don't quite understand why would we want this. Shouldn't we return array of the same type as input? E.g. I expect array_duplicates(array(tinyint)) to return array(tinyint) not array(bigint).

File: velox/functions/prestosql/ArrayDuplicates.cpp [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+        AlignedBuffer::allocate<vector_size_t>(elementsCount, pool);
+    BufferPtr newLengths =
+        AlignedBuffer::allocate<vector_size_t>(rowCount, pool);
+    BufferPtr newOffsets =
+        AlignedBuffer::allocate<vector_size_t>(rowCount, pool);
+
+    // Pointers and cursors to the raw data.
+    vector_size_t indicesCursor = 0;
+    auto* rawNewIndices = newIndices->asMutable<vector_size_t>();
+    auto* rawSizes = newLengths->asMutable<vector_size_t>();
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayDuplicates.cpp:78]: inconsistent naming: rawNewIndices vs. rawSizes; either rename rawNewIndices to rawIndices or rename rawSizes/Offsets to rawNewSizes/Offsets

File: velox/functions/prestosql/ArrayDuplicates.cpp [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+    // Process the rows: use a hashmap to store unique values and
+    // whether it occurred once or more than once.
+    folly::F14FastMap<T, bool> uniqueMap;
+
+    rows.applyToSelected([&](vector_size_t row) {
+      auto size = arrayVector->sizeAt(row);
+      auto offset = arrayVector->offsetAt(row);
+
+      rawOffsets[row] = indicesCursor;
+      vector_size_t numOfNulls = 0;
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayDuplicates.cpp:104]: inconsistent naming: numOfNulls vs. rowCount and elementsCount. Elsewhere in the codebase we use numXxx, e.g. numNulls, numRows and numElements.

File: velox/functions/prestosql/ArrayDuplicates.cpp [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+        }
+      }
+
+      uniqueMap.clear();
+      rawSizes[row] = indicesCursor - rawOffsets[row];
+
+      std::sort(
+          rawNewIndices + rawOffsets[row],
+          rawNewIndices + indicesCursor,
+          sparksql::Index2ValueNullableLess<T>(elements));
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayDuplicates.cpp:143]: sparksql in prestosql directory? Something is not quite right...

File: velox/functions/prestosql/ArrayDuplicates.cpp [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+
+  return VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
+      createTyped, elementType->kind(), inputArgs);
+}
+
+// Define function signature.
+// array(T) -> array(bigint/varchar) where T must be coercible to bigint or
+// varchar.
+std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+  static const std::vector<std::string> kSupportedNumericArgumentTypeNames = {
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayDuplicates.cpp:214]: nit: kSupportedNumericArgumentTypeNames -> kSupportedIntegralTypes

File: velox/functions/prestosql/ArrayDuplicates.cpp (hunk 9 above)
Comment (darrenfu) [reply velox/functions/prestosql/ArrayDuplicates.cpp:120]: Per chatted with @mbasmanova , will limit Velox implementation to inputs of bigint and varchar only for now until we sort out the signature on the Presto side.

File: velox/functions/prestosql/ArrayDuplicates.cpp (hunk 12 above)
Comment (darrenfu) [reply velox/functions/prestosql/ArrayDuplicates.cpp]: Will limit Velox implementation to inputs of bigint and varchar only for now until we sort out the signature on the Presto side

File: velox/docs/functions/array.rst [hunk 17]
```diff
@@ -28,6 +28,13 @@ Array Functions
         SELECT array_distinct(ARRAY [1, 2, 1]); -- [1, 2]
         SELECT array_distinct(ARRAY [1, NULL, NULL]); -- [1, NULL]
 
+.. function:: array_duplicates(array(E)) -> array(E)
+
+    Returns a set of elements that occur more than once in array.
+    E must be coercible to bigint or varchar.
```
Comment (mbasmanova) [inline velox/docs/functions/array.rst:34]: I'd drop "coercible to" since Velox doesn't perform implicit coercions.

File: velox/functions/prestosql/tests/ArrayDuplicatesTest.cpp [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+  void testExpr(
+      const VectorPtr& expected,
+      const std::string& expression,
+      const std::vector<VectorPtr>& input) {
+    auto result = evaluate<ArrayVector>(expression, makeRowVector(input));
+    assertEqualVectors(expected, result);
+  }
+
+  // Execute test for integer types.
+  template <typename T>
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArrayDuplicatesTest.cpp:38]: The test only works with int64_t, hence, no need for a template. I'd also inline this function into integerArrays test and rename the test to bigintArrays.

File: velox/functions/sparksql/Comparisons.h [hunk 19]
```diff
@@ -13,6 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+#include "velox/expression/EvalCtx.h"
```
Comment (mbasmanova) [inline velox/functions/sparksql/Comparisons.h:16]: Why is this header needed?

File: velox/functions/prestosql/ArrayDuplicates.cpp [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+#include "velox/expression/VectorFunction.h"
+#include "velox/functions/lib/ComparatorUtil.h"
+#include "velox/functions/lib/LambdaFunctionUtil.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+// See documentation at https://prestodb.io/docs/current/functions/array.html
+///
+/// This class implements the array_duplicates query function.
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayDuplicates.cpp:30]: - query -> Presto or just drop it
- Drop "This class" for brevity e.g. "Implements the array_duplicates function." or "Implements the array_duplicates Presto function."

File: velox/functions/prestosql/ArrayDuplicates.cpp [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args, // Not using const ref so we can reuse args
+      const TypePtr& outputType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    // Acquire the array elements vector.
+    auto arrayVector = args.front()->as<ArrayVector>();
+    auto elementsVector = arrayVector->elements();
+    ArrayDuplicatesFunction::validateElementType(elementsVector->typeKind());
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayDuplicates.cpp:57]: this doesn't hurt, but is not necessary

File: velox/functions/prestosql/ArrayDuplicates.cpp [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+
+    auto elementsRows =
+        toElementRows(elementsVector->size(), rows, arrayVector);
+    exec::LocalDecodedVector elements(context, *elementsVector, elementsRows);
+
+    vector_size_t numOfElements = elementsRows.size();
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayDuplicates.cpp:63]: the naming convention is numElements and numRows without "Of"

File: velox/functions/prestosql/ArrayDuplicates.cpp [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+    // Prepare and return result set.
+    auto resultArray = std::make_shared<ArrayVector>(
+        pool,
+        outputType,
+        nullptr,
+        numOfRows,
+        std::move(newOffsets),
+        std::move(newSizes),
+        std::move(newElements),
+        0);
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayDuplicates.cpp:131]: remove nulls count; this doesn't quite work and we will be revisiting metadata in vectors

File: velox/functions/prestosql/ArrayDuplicates.cpp [hunk 24]
```diff
@@ -0,0 +1,204 @@
...
+  auto elementType = inputArgs.front().type->childAt(0);
+
+  return VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
+      createTyped, elementType->kind(), inputArgs);
+}
+
+// Define function signature.
+// array(T) -> array(T) where T must be coercible to bigint or varchar.
+std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+  static const std::vector<std::string> kSupportedTypes = {"bigint", "varchar"};
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayDuplicates.cpp:181]: nit: I think it would be easier to read if the signatures are just hard-coded

```
return {
exec::FunctionSignatureBuilder()
            .returnType("array(bigint)")
            .argumentType("array(bigint)")
            .build(),
exec::FunctionSignatureBuilder()
            .returnType("array(varchar)")
            .argumentType("array(varchar)")
            .build()
};
```

**Review by mbasmanova — APPROVED** [review]
@darrenfu Looks great. Just a few minor comments. Please, add an e2e Presto test in a follow-up PR.

File: velox/docs/functions/array.rst (hunk 17 above)
Comment (mbasmanova) [reply velox/docs/functions/array.rst:34]: I expect Presto's query planner to do that. You can test it out by adding an e2e test.

File: velox/functions/prestosql/ArrayDuplicates.cpp (hunk 23 above)
Comment (darrenfu) [reply velox/functions/prestosql/ArrayDuplicates.cpp:130]: Thanks for providing the context!

I agree on overhauling the metadata including nullCount. I think it will be easier to do a one-off fix across all call sites, I'm not over-concerned about the correctness here along with other functions (e.g. array_distinct). Will leave it for now.
